controller/program/src/instructions/liquidity/add.rs

```diff
@@ -0,0 +1,368 @@
+#[inline]
+fn add_liquidity_accs_checked<'a, 'acc>(
+    abr: &Abr,
+    accounts: &'a [AccountHandle<'acc>],
+    ix_args: AddLiquidityIxArgs,
+) -> Result<AddLiquidityIxAccounts<'a, 'acc>, ProgramError> {
+    if ix_args.amount == 0 {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::ZeroValue).into());
+    }
+
+    let (ix_prefix, suf) = accounts
+        .split_first_chunk()
+        .ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let ix_prefix = IxPreAccs(*ix_prefix);
+    let list = LstStatePackedList::of_acc_data(abr.get(*ix_prefix.lst_state_list()).data())
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstStateListData))?;
+
+    // safety: account data is 8-byte aligned
+    let pool = unsafe { PoolState::of_acc_data(abr.get(*ix_prefix.pool_state()).data()) }
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidPoolStateData))?;
+
+    let lst_state = list
+        .0
+        .get(ix_args.lst_index as usize)
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstIndex))?;
+
+    let lst_mint_acc = abr.get(*ix_prefix.lst_mint());
+    let token_prog = lst_mint_acc.owner();
+    // safety: account data is 8-byte aligned
+    let lst_state = unsafe { lst_state.as_lst_state() };
+
+    let expected_reserves =
+        create_raw_pool_reserves_addr(token_prog, &lst_state.mint, &lst_state.pool_reserves_bump)
+            .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_protocol_fee_accumulator = create_raw_protocol_fee_accumulator_addr(
+        token_prog,
+        &lst_state.mint,
+        &lst_state.protocol_fee_accumulator_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_pks = NewAddLiquidityIxPreAccsBuilder::start()
+        // These can be arbitrary accs bc they belong to the user adding liquidity to INF
+        .with_signer(abr.get(*ix_prefix.signer()).key())
+        .with_lst_mint(&lst_state.mint)
+        .with_lst_acc(abr.get(*ix_prefix.lst_acc()).key())
+        .with_lp_acc(abr.get(*ix_prefix.lp_acc()).key())
+        .with_lp_token_mint(&pool.lp_token_mint)
+        .with_protocol_fee_accumulator(&expected_protocol_fee_accumulator)
+        .with_lst_token_program(&TOKEN_PROGRAM)
+        .with_lp_token_program(&TOKEN_PROGRAM)
+        .with_pool_state(&POOL_STATE_ID)
+        .with_lst_state_list(&LST_STATE_LIST_ID)
+        .with_pool_reserves(&expected_reserves)
+        .build();
+
+    verify_pks(abr, &ix_prefix.0, &expected_pks.0)?;
+
+    let (lst_cal_all, pricing_all) = suf
+        // Adding +1 here bc we need to take into account the program as well
+        .split_at_checked((ix_args.lst_value_calc_accs + 1).into())
+        .ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let (lst_calc_prog, lst_calc_acc) = lst_cal_all.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let (pricing_prog, pricing_accs) = pricing_all.split_first().ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
```

Comment:
nit: array.map + let-else can be used here
https://github.com/igneous-labs/inf-1.5/pull/77/files#diff-f4d8d5460e8f5b75141b3d26701c3b7c76fbe57e760b08e5d43cb22386b18d53R136-R140